std/int128.d

```diff
@@ -0,0 +1,290 @@
+// Written in the D programming language
+/**
+ * Implements a signed 128 bit integer type.
+ *
+    Author:     Walter Bright
+    Copyright:  Copyright (c) 2022, D Language Foundation
+    License:    $(HTTP boost.org/LICENSE_1_0.txt, Boost License 1.0)
+    Source:     $(PHOBOSSRC std/int128.d)
+ */
+module std.int128;
+
+private import core.int128;
+
+
+/***********************************
+ * 128 bit signed integer type.
+ */
+
+public struct Int128
+{
+  @safe pure nothrow @nogc:
+
+    Cent data;
+
+    /****************
+     * Construct an `Int128` from a `long` value.
+     * The upper 64 bits are formed by sign extension.
+     * Params:
+     *  lo = signed lower 64 bits
+     */
+    this(long lo)
+    {
+        data.lo = lo;
+        data.hi = lo < 0 ? ~0L : 0;
+    }
+
+    /****************
+     * Construct an `Int128` from a `long` value.
+     * Params:
+     *  hi = upper 64 bits
+     *  lo = lower 64 bits
+     */
+    this(long hi, long lo)
+    {
+        data.hi = hi;
+        data.lo = lo;
+    }
+
+    this(Cent data)
+    {
+        this.data = data;
+    }
+
+    bool opEquals(long lo) const
+    {
+        return data.hi == 0 && data.lo == lo;
+    }
+
+    bool opEquals(Int128 op2) const
+    {
+        return data.hi == op2.data.hi && data.lo == op2.data.lo;
+    }
+
+    // +Int128
+    Int128 opUnary(string op)() const
+        if (op == "+")
+    {
+        return this;
+    }
+
+    // -Int128
+    Int128 opUnary(string op)() const
+        if (op == "-" || op == "~")
+    {
+        static if (op == "-")
+            return Int128(neg(this.data));
+        else static if (op == "~")
+            return Int128(com(this.data));
+    }
+
+    Int128 opUnary(string op)()
+        if (op == "++" || op == "--")
+    {
+        static if (op == "++")
+            this.data = inc(this.data);
+        else static if (op == "--")
+            this.data = dec(this.data);
+        else
+            static assert(0, op);
+        return this;
+    }
+
+    // cast(T)Int128
+    bool opCast(T : bool)() const
+    {
+        return tst(this.data);
+    }
+
+    // Int128 + Int128
+    Int128 opBinary(string op)(Int128 op2) const
+        if (op == "+" || op == "-" ||
+            op == "*" || op == "/" || op == "%" ||
+            op == "&" || op == "|" || op == "^")
+    {
+        static if (op == "+")
```

Comment:
Why we are doing template specialization by hand?

Reply:
Looks straightforward to me.

Reply:
For me, looks more readable to find the correct specialization with something like this:

```d
Int128 opBinary(string op : "+")(Int128 op2) const
{
    return Int128(add(this.data, op2.data));
}
```

Shouldn't be a big deal but, also, specializing in this way would be faster for the compiler to look it up, as opposed to constraints, in which it needs to evaluate the static condition, in this case, twice.

Reply:
My way is easier to read (for me, anyway) as it emphasizes the commonality.

Reply:
I'd also prefer the constraint in the header (whether by `if` or `:` isn't that important) to avoid duplication, but I allow your style if you prefer it.